sphinx/util/inspect.py

```diff
@@ -110,12 +110,23 @@ def safe_getattr(obj, name, *defargs):
     except Exception:
         # sometimes accessing a property raises an exception (e.g.
         # NotImplementedError), so let's try to read the attribute directly
-        if name in obj.__dict__:
+
+        # We should also be aware that if `__dict__` has been overridden,
+        # this may also raise an exception.
+
+        try:
+            obj_dict = obj.__dict__
+        except Exception:
+            obj_dict = {}
```

Comment:
I wanted to be very specific about where I was passing the exception, as it's such a broad `except` clause. Perhaps this is clearer:

```python
        try:
            # In case __dict__ has been redefined as a `@property` such that
            # it may raise an error on access
            obj_dict = obj.__dict__
        except Exception:
            pass
        else:
            # We don't want to catch any exceptions here
            if name in obj_dict:
                return obj_dict[name]
```